File: 卡码/代码随想录/Python/动态规划/05-子序列问题/04-回文/LeetCode_0647.py
#647. 回文子串
from typing import *

class Solution:
  #动态规划
  def countSubstrings(self, s: str) -> int:
    dp:List[List[bool]]=[[False]*len(s) for _ in range(len(s))]
    res:int=0
    for i in range(len(s)-1,-1,-1):
      for j in range(i,len(s)):
        if s[i]==s[j] and (j-i<=1 or dp[i+1][j-1]):
          dp[i][j]=True
          res+=1
    return res
  #双指针
  def countSubstrings1(self, s: str) -> int:
    res:int=0
    for i in range(len(s)):
      res+=self.countPalindrome(s, i, i)
      res+=self.countPalindrome(s, i, i+1)
    return res
  def countPalindrome(self,s:str,i:int,j:int)->int:
    res:int=0
    while i>=0 and j<len(s) and s[i]==s[j]:
      i-=1
      j+=1
      res+=1
    return res

File: 卡码/代码随想录/Go/动态规划/05-子序列问题/04-回文/LeetCode_0647.go
// 647. 回文子串
package main

// 动态规划
func countSubstrings(s string) int {
	res := 0
	dp := make([][]bool, len(s))
	for i := 0; i < len(s); i++ {
		dp[i] = make([]bool, len(s))
	}
	for i := len(s) - 1; i >= 0; i-- {
		for j := i; j < len(s); j++ {
			if s[i] == s[j] && (j-i <= 1 || dp[i+1][j-1] == true) {
				dp[i][j] = true
				res++
			}
		}
	}
	return res
}

// 双指针
func countSubstrings1(s string) int {
	res := 0
	for i := 0; i < len(s); i++ {
		res += countPalindrome(s, i, i)
		res += countPalindrome(s, i, i+1)
	}
	return res
}
func countPalindrome(s string, i, j int) int {
	res := 0
	for i >= 0 && j < len(s) && s[i] == s[j] {
		i--
		j++
		res++
	}
	return res
}
